### ISSUE TITLE
Design discussion

### ISSUE THREAD

**Issue by phipsgabler** [phipsgabler/DensityInterfaceAD.jl#1]
Opening this to discuss what I wrote. The basic idea is to have a new wrapper type, `DifferentiableLogFuncDensity`, that works the same as `LogDensityFunc` (and `IsDensity`), but also conforms to the LogDensityProblems interface and implements `logdensity`, `dimension`, and `capabilities`. The wrapper is really just there to hold the additionally needed `dimension` information.

This object can then just be wrapped in a `LogDensityProblems.ADGradient`. The `ADGradient` itself now also should have `logdensityof` and `IsDensity` (which is type piracy; ideally, these methods should live in LogDensityProblems directly).

Finally, the easier way to get all this is a new method `logfuncdensity(f, ::ADConfig)` which does all the wrapping.

Any suggestions?

@oschulz If this is approved to make sense, feel free to move under JuliaMath to live with DensityInterface.

**Comment by oschulz** [comment]
Regarding the design - maybe we should choose a different name for `DifferentiableLogFuncDensity`? Currently, it's more a "density with dimension", there's nothing differentiation-specific about it, right?

Regarding dependencies - DensityInterface.jl is very lightweight, wouldn't it make more sense for LogDensityProblems.jl to depend on DensityInterface.jl and DensityInterfaceAD.jl, instead of having DensityInterfaceAD.jl depend on LogDensityProblems.jl?

> @oschulz If this is approved to make sense, feel free to move under JuliaMath to live with DensityInterface.

I'm not an owner of JuliaMath, so I should still talk to one of JuliaMath "regulars", to be polite (we should finish the package first, probably).

**Comment by oschulz** [comment]
Question regarding dimension - which use cases do we need the dimension information, and how? And what's the exact meaning, is it degrees of freedom?

@devmotion, maybe we could "upstream" an optional "query-dimension" function to DensityInterface.jl? It this is about actual DOF - we already have `getdof(x)` in MeasureBase.jl (CC @cscherrer), maybe we could "upstream" that to DensityInterface.jl?

**Comment by devmotion** [comment]
AFAICT it's not about DOF. It's just a way to retrieve the length of the input vector *before* you've seen any actual input since this is needed for e.g. optimized gradient configurations with ForwardDiff and tape compilations with ReverseDiff.

Generally, the design of this package here is quite unclear to me though. I think https://github.com/phipsgabler/DensityInterfaceAD.jl/blob/31c22971c696291cb8861c34e186949623d42995/src/DensityInterfaceAD.jl#L50-L69 should all not be in this package ideally, the first few due to type piracy and AD types are much more generally useful (I thought about adding something similar but specialized for each backend (to avoid a generic `kwargs`) to LogDensityProblems as that could be reused e.g. in Turing). 
I'm also not sure about the best dependency structure. LogDensityProblems will become more and more lightweight, `TransformVariables` will be gone soon and in the long run maybe AD could use AbstractDifferentiation at least partly.

One also does not always need a separate wrapper if one works with LogDensityProblems and, possibly, AD: for instance, if one uses a transformed log density function (that's about to be extracted into its own package but so far has been the recommended way to define something compatible with the interface), then the dimension is retrieved from the transformation but not the log density function. In fact, in this case LogDensityProblems already supports arbitrary functions as log density functions, e.g., one can just use `logdensityof(whatever)`.

In general, my feeling is that usually one knows if one operates with and expects some density that should be evaluated (for which the DensityInterface is sufficient) or if one needs additional support e.g. for compiled tapes, AD specializations etc. (in which case either one implements the desired features or uses something like LogDensityProblems). So I guess one usually knows if DensityInterface is sufficient or one needs a different interface, and hence one can just support the most appropriate one and there's no need to switch between those? Maybe it's sufficient to just define `capabilities`, `dimension`, and `logdensity` for a custom density when it's needed - without any wrappers or additional packages? The transformed case already works automatically, and in the other cases one would have to define them anyway, also when not using something compatible with DensityInterface.

**Comment by oschulz** [comment]
> and logdensity for a custom density when it's needed

`logdensity` essentiall duplicates `logdensityof` - maybe LogDensityProblems could adopt it, resp. make `LogDensityProblems.logdensity` a synonym for `DensityInterface.logdensityof` (CC @tpapp)?

**Comment by tpapp** [comment]
> `logdensity` essentiall duplicates `logdensityof`

I think it is the other way around, since LogDensityProblems predates DensityInterface, so if anything, the latter duplicates the former. I do not understand the logic behind duplicating an interface in another package *and* then asking that the original package adds a synonym for the new package.

As I stated multiple times, LogDensityProblems will remove dependency on TransformVariables (that will happen in a few days) and then depend on AbstractDifferentiation when the latter is mature (current `LogDensityProblems.ADgradient` is a workaround).

Personally I see no compelling reason for another package. It would be great if the discussion could clarify what is missing in LogDensityProblems and then maybe can just add it? (@devmotion now has commit rights)

**Comment by tpapp** [comment]
> @devmotion It's just a way to retrieve the length of the input vector before you've seen any actual input since this is needed for e.g. optimized gradient configurations with ForwardDiff

Yes, and it is also needed for a lot of other things (generate random starting points for MCMC, etc).

**Comment by oschulz** [comment]
>> logdensity essentiall duplicates logdensityof
> I think it is the other way around, since LogDensityProblems predates DensityInterface, so if anything, the latter duplicates the former. I do not understand the logic behind duplicating an interface in another package and then asking that the original package adds a synonym for the new package.

In general, one should avoid duplicating existing interfaces, of course. In this specific case, however, DensityInterface was created for the express purpose of creating an ultra-lightweight community-wide interface for densities/measures/distributions. The interface was the result of extensive discussions of all of us (you were involved as well, if I remember correctly, @tpapp?). In a case like this, where a package is more lightweight and designed as a community API, I would say it is reasonably to suggest that pre-existing packages in the domain adopt/support the new API, especially if this can be done without breaking their backwards compatibility.
